File: SC/volume2.py
# transcription of volume2.m
import matplotlib.pyplot as plt
from skimage.color import rgb2gray
from scipy import ndimage
from skimage.feature import canny
from skimage.transform import hough_line, hough_line_peaks
import numpy as np
import cv2 as cv

# # read image
# image_rgb = plt.imread("picture_1.png") # values between 0 and 1, float, 3d image (h, w, channel)

# # convert to gray image
# image_gray = rgb2gray(image_rgb) # values between 0 and 1, float, 2d image (h, w)
# # plt.imshow(image_gray, cmap="gray")

# # convert grayscale to boolean field of intensity larger than given factor
# Ibinary = image_gray > 120 / 255
# # plt.imshow(Ibinary, cmap="gray")

# # fill holes
# Ibinary = ndimage.binary_fill_holes(Ibinary).astype(int)
# # plt.imshow(Ibinary, , cmap="gray")

# # detect edges in image using edge-detection algorithm "canny"
# # https://scikit-image.org/docs/dev/api/skimage.feature.html#skimage.feature.canny
# BW = canny(Ibinary*1.0) # binary edge map
# plt.imshow(BW)

# # # perform a straight line Hough transform, returns hough transform accumulator, angles and distance
# # # https://scikit-image.org/docs/dev/api/skimage.transform.html?highlight=hough%20circle#skimage.transform.hough_line
# # H, theta, rho  = hough_line(BW)

# # # return peaks in a straight line Hough transform. Identifies most prominent lines separated by a certain angle and distance in a Hough transform. 
# # # https://scikit-image.org/docs/dev/api/skimage.transform.html?highlight=hough%20circle#skimage.transform.hough_line_peaks
# # H, theta, rho = hough_line_peaks(H, theta, rho, threshold=np.ceil(0.5*max(H)), num_peaks=6)
# # lines = HoughLines(BW)


img = cv.imread(cv.samples.findFile('picture_1.png'))
gray = cv.cvtColor(img,cv.COLOR_BGR2GRAY) # between 0 and 255
print(gray)
# gray = (gray>120)*255.0
gray = (gray>120)

print(gray)

# # fill holes
Ibinary = np.uint8(ndimage.binary_fill_holes(gray))*255
# # plt.imshow(Ibinary, , cmap="gray")

print(Ibinary)

edges = cv.Canny(Ibinary,0,255,apertureSize = 3)

# while(True):
# 	cv.imshow('frame', edges)
# 	if cv.waitKey(1) & 0xFF == ord('q'):
# 		break
average=0
lines = cv.HoughLinesP(edges,1,np.pi/360,98,minLineLength=7,maxLineGap=5)
for line in lines:
    x1,y1,x2,y2 = line[0]
    average+= (y1+y2)/2.0

print(average/len(lines))

File: SC/volume2.m
pkg load image
I = imread('picture_1.png'); %read image

I = rgb2gray(I) > 120; % Converts image to greyscale by analizing luminance. Try different values for saturation calibration
I = imfill(I, 'holes'); % fills holes in the input binary image BW. In this syntax, a hole is a set of background pixels that cannot be reached by filling in the background from the edge of the image.
#imshow(I) % show greyscale image

BW = edge(I*1.0,'canny'); %detects edges in image I using the edge-detection algorithm "canny". The Canny method finds edges by looking for the local maxima of the gradient of the input image.
imshow(BW)
[H,theta,rho] = hough(BW); %Hough transform: matrix of line parameters. A line is represented by rho= x*cos(theta) + y*sin(theta). 
%The function returns rho, the distance from the origin to the line along a vector perpendicular to the line, 
%and theta, the angle in degrees between the x-axis and this vector. The function also returns the Standard Hough Transform, H, 
%which is a parameter space matrix whose rows and columns correspond to rho and theta values respectively.

P = houghpeaks(H,6,'threshold',ceil(0.5*max(H(:)))); %identify maximum 5 peaks in Hough transform matrix H. 
%The function returns peaks a matrix that holds the row and column coordinates of the peaks.
%'threshold' is the name assigned to the minimum value (0.3*maximum value of H) to be considered a peak 

lines = houghlines(BW,theta,rho,P,'FillGap',5,'MinLength',7); %extracts line segments in the image BW. theta and rho are vectors returned by hough peaks.
%Distance between two line segments associated with the same Hough transform bin, specified as a positive number. 
%FillGap: When the distance between the line segments is less than the value specified (5), the houghlines function merges the line segments into a single line segment.
%MinLength: Minimum line length, specified as a positive number. houghlines discards lines that are shorter than the value specified (7).
%lines is a vector
%disp(lines)

average=0;
for k = 1:length(lines)
  xy = [lines(k).point1; lines(k).point2];
  average=average + (xy(1,2)+xy(2,2))/2;
end

disp(average/length(lines));
